### PR TITLE
Week1 cleanup

### PR DESCRIPTION
**Pull request by matthias-epos**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,17 +4,20 @@ This project is used to test different approaches to storing information with ch
...
 
-Currently two different approaches are being investigated:
+Currently, two different approaches are being investigated:
 
 ## EAV
 
-The entity-attribute-value model disregards the intended usage of relations to create a dynamic model. It consists of three tables: Entity, Attribute and Values. In Entity and Attribute we save the definition of entities and attributes respectively. The values table saves in each row one value of an attribute of an entity.
+The entity-attribute-value model disregards the intended usage of relations to create a dynamic model. 
+It consists of three tables: Entity, Attribute and Values. In Entity and Attribute we save the definition of entities and attributes respectively. 
+The values table saves in each row one value of an attribute of an entity.
```
Comment (milena-i) [inline README.md]:
```suggestion
The **Values** table stores in each row one value of an attribute of an entity.
```

File: README.md [hunk 2]
```diff
@@ -4,17 +4,20 @@ This project is used to test different approaches to storing information with ch
...
 We want to be able to add custom attribute-value pairs to entities during runtime of the program.
 
-Currently two different approaches are being investigated:
+Currently, two different approaches are being investigated:
 
 ## EAV
 
-The entity-attribute-value model disregards the intended usage of relations to create a dynamic model. It consists of three tables: Entity, Attribute and Values. In Entity and Attribute we save the definition of entities and attributes respectively. The values table saves in each row one value of an attribute of an entity.
+The entity-attribute-value model disregards the intended usage of relations to create a dynamic model. 
+It consists of three tables: Entity, Attribute and Values. In Entity and Attribute we save the definition of entities and attributes respectively.
```
Comment (milena-i) [inline README.md]:
```suggestion
It consists of three tables: **Entity, Attribute and Values**. In Entity and Attribute we save the definition of entities and attributes respectively. 
```

File: README.md [hunk 3]
```diff
@@ -4,17 +4,20 @@ This project is used to test different approaches to storing information with ch
 
 We want to be able to add custom attribute-value pairs to entities during runtime of the program.
 
-Currently two different approaches are being investigated:
+Currently, two different approaches are being investigated:
 
 ## EAV
 
-The entity-attribute-value model disregards the intended usage of relations to create a dynamic model. It consists of three tables: Entity, Attribute and Values. In Entity and Attribute we save the definition of entities and attributes respectively. The values table saves in each row one value of an attribute of an entity.
+The entity-attribute-value model disregards the intended usage of relations to create a dynamic model.
```
Comment (milena-i) [inline README.md:11]: Is the usage of _diregards_ intentional? It means _ignores_.

File: README.md [hunk 4]
```diff
@@ -4,17 +4,20 @@ This project is used to test different approaches to storing information with ch
...
-The entity-attribute-value model disregards the intended usage of relations to create a dynamic model. It consists of three tables: Entity, Attribute and Values. In Entity and Attribute we save the definition of entities and attributes respectively. The values table saves in each row one value of an attribute of an entity.
+The entity-attribute-value model disregards the intended usage of relations to create a dynamic model. 
+It consists of three tables: Entity, Attribute and Values. In Entity and Attribute we save the definition of entities and attributes respectively. 
+The values table saves in each row one value of an attribute of an entity.
 
 ## JSON-Values
 
-JSON enables us to store dynamic structures structures in a text format. Thanks to the increased usage of JSON in web environments database vendors improved their handling of this data type.
+JSON enables us to store dynamic structures in a text format. 
+Thanks to the increased usage of JSON in web environments database vendors improved their handling of this data type.
```
Comment (milena-i) [inline README.md]:
```suggestion
Thanks to the increased usage of JSON in web environments database vendors improved the handling of this data type.
```

File: master.iml [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (milena-i) [inline master.iml]: This file should not be checked-in. Please add *.iml files to git ignore.

File: pom.xml [hunk 6]
```diff
@@ -5,34 +5,75 @@
   <modelVersion>4.0.0</modelVersion>
 
   <groupId>de.eposcat</groupId>
-  <artifactId>master</artifactId>
+  <artifactId>dynamic-datamodels</artifactId>
   <version>0.0.1-SNAPSHOT</version>
 
-  <name>master</name>
+  <name>Dynamic Data Model</name>
   <!-- FIXME change it to the project's website -->
```
Comment (milena-i) [inline pom.xml]: The _FIXME_ comment can be remove.

File: src/main/java/de/eposcat/master/App.java [hunk 7]
```diff
@@ -1,66 +1,57 @@
...
 import de.eposcat.master.connection.AbstractConnectionManager;
 import de.eposcat.master.connection.CustomOracleConnectionManager;
 import de.eposcat.master.connection.RelationalApproach;
-import de.eposcat.master.model.Attribute;
+import de.eposcat.master.model.AttributeBuilder;
 import de.eposcat.master.model.AttributeType;
 import de.eposcat.master.model.Page;
 
 /**
  * Hello world!
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/App.java]:
```suggestion
 * The main class of the Application
```

File: src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java [hunk 8]
```diff
@@ -6,150 +6,170 @@
...
-        
-        return null;
+
+        if (updatedRows > 0) {
+            //TODO? Cancel Transaction if getting generated keys fails?
+            ResultSet keySet = st.getGeneratedKeys();
+            return new Page(getIdFromGeneratedKeys(keySet, ENTITY_TABLE), typeName);
+        }
+
+        throw new SQLException("No new row has been created in database. This should never happen!");
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java]:
```suggestion
        throw new SQLException("Could not insert a new page in the entity table. This should never happen!");
```

File: src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java [hunk 9]
```diff
@@ -6,150 +6,170 @@
...
+    private final static String ENTITY_TABLE =  "entityTable";
+    private final static String ATTRIBUTE_TABLE = "attributeTable";
+
     public EAV_DatabaseAdapter(AbstractConnectionManager connectionManager) {
         this.conn = connectionManager.getConnection();
     }
 
     @Override
-    public Page createPage(String typeName) throws SQLException{
+    public Page createPage(String typeName) throws SQLException {
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java:48]:
```suggestion
    @Nonnull
    public Page createPage(String typeName) throws SQLException {
```
If this method always returns a non-null object, you could annotate it with @Nonnull (javax.annotation.Nonnull)

File: src/main/java/de/eposcat/master/exceptions/BlException.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package de.eposcat.master.exceptions;
+
+public class BlException extends RuntimeException{
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/exceptions/BlException.java:6]: Missing javadoc. What is the abbreviation **Bl**?

File: src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java [hunk 11]
```diff
@@ -6,150 +6,170 @@
...
             PreparedStatement st = conn.prepareStatement(UPDATE_PAGE_QUERY);
             st.setString(1, page.getTypeName());
-            st.setInt(2, page.getId());
+            st.setLong(2, page.getId());
             int affectedRows = st.executeUpdate();
-            
+
             if(affectedRows != 1) {
-                throw new SQLException();
+                throw new BlException("page is not tracked by database, try create pageWithAttributes first");
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java]: The pageId (or some page info) would be useful in the exception message, e.g.
```suggestion
                throw new BlException("Page with id=xyz, type=xyz is not tracked by database, try create pageWithAttributes first");
```

File: src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java [hunk 12]
```diff
@@ -6,150 +6,170 @@
...
-            });
-            
+            }
+
+            for(long id : currentAttributes){
+                removeAttributeValue(page.getId(), id);
+            }
+
             conn.commit();
         } finally {
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java:124]: Why do you need **conn.setAutoCommit(true);** in the finally block?

File: src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java [hunk 13]
```diff
@@ -161,79 +181,100 @@ private Attribute saveAttribute(String attributeName, Attribute attribute) throw
...
+        while(rsPages.next()) {
+            pageList.add(loadPage(rsPages.getInt(1)));
+        }
+
+        return pageList;
+    }
+
     //Add attributeType?
     @Override
     public List<Page> findPagesByAttribute(String attributeName) throws SQLException{
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java]: If this method searches only by attribute name, then the name could be more precise: **findPagesByAttributeName**

File: src/main/java/de/eposcat/master/approachImpl/IDatabaseAdapter.java [hunk 14]
```diff
@@ -2,15 +2,19 @@
...
 
 public interface IDatabaseAdapter {
-    public Page createPage(String typename) throws SQLException;
-    
-    public void updatePage(Page page) throws SQLException;
-    public Page loadPage(int pageId) throws SQLException;
-    
-    public List<Page> findPagesByAttribute(String AttributeName) throws SQLException;
-    public List<Page> findPagesByAttributeValue(String AttributeName, Object value) throws SQLException;
+    Page createPage(String typename) throws SQLException;
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/approachImpl/IDatabaseAdapter.java:12]: Missing javadoc on the interface methods. Could be added in a separate PR.

File: src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java [hunk 15]
```diff
@@ -82,6 +91,27 @@ public Page loadPage(int pageId) throws SQLException{
         }
     }
 
+    @Override
+    public List<Page> findPagesByType(String type) throws SQLException {
+        PreparedStatement stFindByType = conn.prepareStatement("SELECT * FROM pages WHERE type = ?");
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java:96]: Missing null-check for the type parameter

File: src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java [hunk 16]
```diff
@@ -82,6 +91,27 @@ public Page loadPage(int pageId) throws SQLException{
...
+            page.setAttributes(jsonToMap(attributesJSON));
+            pages.add(page);
+        }
+
+        return pages;
+    }
+
     @Override
     public List<Page> findPagesByAttribute(String attributeName) throws SQLException{
         // Oracle String literals are stupid -> sqlInjection would be possible here...
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java:117]: This should be documented in the javadoc for this method, maybe with links to some examples/explanations.

File: src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java [hunk 17]
```diff
@@ -20,54 +21,62 @@
...
     }
 
     @Override
     public Page createPage(String typename) throws SQLException{
+        return createPageWithAttributes(typename, new HashMap<>());
+    }
+
+    @Override
+public Page createPageWithAttributes(String typename, Map<String, Attribute> attributes) throws SQLException{
         PreparedStatement stCreatePage = conn.prepareStatement("INSERT INTO pages(type, attributes) VALUES (?,?)", Statement.RETURN_GENERATED_KEYS);
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java:39]: Missing null-checks for the parameters.

File: src/main/java/de/eposcat/master/model/Page.java [hunk 18]
```diff
@@ -2,30 +2,32 @@
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
 
 public class Page {
-    private int id;
+    private long id;
+    //TODO make typeName mandatory
```
Comment (milena-i) [inline src/main/java/de/eposcat/master/model/Page.java:9]: Not sure, currently the type is not used as a grouping for attributes. One option would be to remove it because it is not used. Another would be to rename it to pageName, which seems to be more in-line with the current implementation. This TODO: should be tackled in a separate PR.

File: src/main/resources/sql/postgre.sql [hunk 19]
```diff
@@ -0,0 +1,52 @@
+-- Currently the Connection URL for the connector is jdbc:postgresql://localhost/ + database name (based on approach used)
+-- We should add a properties or config file to change it on the fly
```
Comment (milena-i) [inline src/main/resources/sql/postgre.sql:2]: Adding a system property is a good idea. This should be addressed in a separate PR.

File: src/test/java/de/eposcat/master/AppTest.java [hunk 20]
```diff
@@ -1,8 +1,9 @@
 package de.eposcat.master;
 
-import static org.junit.Assert.assertTrue;
 
-import org.junit.Test;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
Comment (milena-i) [inline src/test/java/de/eposcat/master/AppTest.java]: Maybe this class should be removed as it does not test any business logic.

File: src/test/java/de/eposcat/master/PageTest.java [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    @Test
+    public void createPage() {
+        try {
+            //returned Pages must have an id, the given type and no attributes
+            Page page = dbAdapter.createPage("testType");
+            assertThat(page.getId(), not(-1));
+            assertThat(page.getTypeName(), is("testType"));
+            assertThat(page.getAttributes().isEmpty(), is(true));
```
Comment (milena-i) [inline src/test/java/de/eposcat/master/PageTest.java:46]: Very good 👍

File: src/test/java/de/eposcat/master/PageTest.java [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+
+            //fails on  missing/empty type
+            assertThrows(IllegalArgumentException.class, () -> dbAdapter.createPage(null));
+            assertThrows(IllegalArgumentException.class, () -> dbAdapter.createPage(""));
```
Comment (milena-i) [inline src/test/java/de/eposcat/master/PageTest.java]: These 2 checks can be added in a separate @Test method, e.g. testCreatePageWithInvalidType

File: src/test/java/de/eposcat/master/PageTest.java [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+        } catch (SQLException e) {
+            fail();
+            e.printStackTrace();
+        }
+
+
+
+    }
+
+    //TODO rename to findByAttributeName
```
Comment (milena-i) [inline src/test/java/de/eposcat/master/PageTest.java]: Ha, I had the same suggestion :) :)

File: src/test/java/de/eposcat/master/PageTest.java [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+import java.util.HashMap;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.fail;
+
+
+public class PageTest {
```
Comment (milena-i) [inline src/test/java/de/eposcat/master/PageTest.java:26]: The tests look good in general but it seems that multiple test cases are covered with one method.
Some test cases could be grouped together and tested in a separate method in order to increate readability and to reduce the complexity of the tests.

File: src/test/java/de/eposcat/master/PageTest.java [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+
+    static IDatabaseAdapter dbAdapter;
+    static Attribute defaultAttribute;
+
+    @BeforeAll
+    static void initDataBase() {
+        H2ConnectionManager connectionManager = new H2ConnectionManager(RelationalApproach.EAV);
```
Comment (milena-i) [inline src/test/java/de/eposcat/master/PageTest.java:33]: This database setup will be needed in other test classes and can be extracted in a super **Test** class to avoid code duplication. This refactoring can be done a separate PR.

**Review by milena-i — CHANGES_REQUESTED** [review]

File: src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java (hunk 9 above)
Comment (matthias-epos) [reply src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java:48]: Is there a certain dependency I need to use to get @Nonnull?

My IDE does not find the annotation at the moment, only @NotNull.

File: src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java (hunk 12 above)
Comment (matthias-epos) [reply src/main/java/de/eposcat/master/approachImpl/EAV_DatabaseAdapter.java:124]: We disable auto commit at the beginning of this method, since it is a multi transaction process to persist a single page in the database. There should be rollback if one of the transactions fails. After finishing persisting the page we want to use auto commit again, even if an error occurred. 

The other methods use auto commit since it is only a single transaction for every method.

File: README.md (hunk 3 above)
Comment (matthias-epos) [reply README.md:11]: Yes it is. I might rewrite this sentence or extend this section to make my intentions more clear.

File: src/main/java/de/eposcat/master/exceptions/BlException.java (hunk 10 above)
Comment (matthias-epos) [reply src/main/java/de/eposcat/master/exceptions/BlException.java:6]: bl is short for business logic. Not sure if it is the best term.

File: src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java (hunk 15 above)
Comment (matthias-epos) [reply src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java:96]: I'll rework both JSON adapters once the testing environments are ready.

**Review by milena-i — APPROVED** [review]

File: src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java (hunk 16 above)
Comment (matthias-epos) [reply src/main/java/de/eposcat/master/approachImpl/JSON_Oracle_DatabaseAdapter.java:117]: #3 I've added the Javadoc in another PR.
